rpcs3/Emu/RSX/RSXFIFO.cpp

```diff
@@ -58,16 +66,115 @@ namespace rsx
 			}
 		}
 
-		void FIFO_control::set_get(u32 get, bool check_spin)
+		std::pair<bool, u32> FIFO_control::fetch_u32(u32 addr)
 		{
-			if (check_spin && m_ctrl->get == get)
+			if (addr - m_cache_addr >= m_cache_size)
 			{
-				if (const u32 addr = m_iotable->get_addr(m_memwatch_addr); addr + 1)
+				const u32 put = read_put();
+
+				if (put == addr)
 				{
-					m_memwatch_addr = get;
-					m_memwatch_cmp = vm::read32(addr);
+					return {false, FIFO_EMPTY};
 				}
 
+				m_cache_addr = addr & -128;
+
+				const u32 addr1 = m_iotable->get_addr(m_cache_addr);
+
+				if (addr1 == umax)
+				{
+					m_cache_size = 0;
+					return {false, FIFO_ERROR};
+				}
+
+				if (!g_cfg.core.rsx_accurate_fifo_access)
+				{
+					std::memcpy(&m_cache, vm::base(addr1 + (addr & 0x60)), 32);
```

Comment:
We know some things here that the compiler cannot know and that is an optimization opportunity. The addresses are guaranteed to be aligned, we can do this in 1 avx or 2 SSE instructions easy.